### ISSUE TITLE
[css-grid][css-flexbox][quirks] Avoid percentage height quirk in new layout models

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#5545]
Quirks mode has this [percentage quirk](https://quirks.spec.whatwg.org/#the-percentage-height-calculation-quirk) which skips ancestors with `height: auto` when resolving `height: <percentage>`. Basically, it's resolved against the nearest ancestor in the containing block chain that doesn't have `height: auto`.

There are some exceptions: grid and flex *containers* are not skipped.

But we still have an interoperability problem: Firefox doesn't skip grid and flex *items*, while Chromium does. [Testcase](https://software.hixie.ch/utilities/js/live-dom-viewer/?%3Cdiv%20style%3D%22display%3A%20flex%3B%20height%3A%20100px%3B%20width%3A%20max-content%3B%20border%3A%205px%20solid%20blue%22%3E%0A%20%20%3Cdiv%20style%3D%22width%3A%20100px%3B%20align-self%3A%20center%3B%20border%3A%203px%20solid%20lime%22%3E%0A%20%20%20%20%3Cdiv%20style%3D%22height%3A%20100%25%3B%20background%3A%20red%22%3E%3C%2Fdiv%3E%0A%20%20%3C%2Fdiv%3E%0A%3C%2Fdiv%3E%0A%3Cdiv%20style%3D%22display%3A%20grid%3B%20grid-template-rows%3A%20100px%3B%20width%3A%20max-content%3B%20border%3A%205px%20solid%20blue%22%3E%0A%20%20%3Cdiv%20style%3D%22width%3A%20100px%3B%20align-self%3A%20center%3B%20border%3A%203px%20solid%20lime%22%3E%0A%20%20%20%20%3Cdiv%20style%3D%22height%3A%20100%25%3B%20background%3A%20red%22%3E%3C%2Fdiv%3E%0A%20%20%3C%2Fdiv%3E%0A%3C%2Fdiv%3E)

```html
<div style="display: flex; height: 100px; width: max-content; border: 5px solid blue">
  <div style="width: 100px; align-self: center; border: 3px solid lime">
    <div style="height: 100%; background: red"></div>
  </div>
</div>
<div style="display: grid; grid-template-rows: 100px; width: max-content; border: 5px solid blue">
  <div style="width: 100px; align-self: center; border: 3px solid lime">
    <div style="height: 100%; background: red"></div>
  </div>
</div>
```

![screenshot](https://user-images.githubusercontent.com/7477678/94060291-01099f00-fde4-11ea-9c7a-a2e80117890c.png)

I think Firefox is right and the quirk should be avoided for elements that participate in new layout models (like grid-level and flex-level boxes).

It's not just that quirks shouldn't infect new features. There is also the thing that stretched grid/flex items are considered to have a definite height, even if they have `height: auto`. So skipping them in that case doesn't even make sense. And skipping them depending on whether the height will end up being definite or not would just make it harder for implementations.

**Comment by dholbert** [comment]
For reference, I implemented Firefox's behavior on this scenario in https://bugzilla.mozilla.org/show_bug.cgi?id=1578586 .

**Comment by chrishtr** [comment]
@bfgeek

**Comment by Loirooriol** [comment]
BTW, I said "would just make it harder for implementations" because in Chromium I tried checking `HasPercentHeightDescendants()`, but it returned false if the grid item was skipped. Turns out that `MaybeHasPercentHeightDescendant()` did the trick. So this point may not hold, but I would still prefer to avoid the quirk in new layout models.

**Comment by bfgeek** [comment]
Due to flexbox and definiteness changing (with flexing/stretching) you need to perform very precise tracking of %-block-size descendants as an arbitrary descendant may trigger a relayout. E.g.
https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=8709

**Comment by dholbert** [comment]
> RESOLVED: dholbert to make a change to quirks spec to clarify behavior for flex and grid items and containers

Looking at the quirks spec-text more clearly, I agree that it does need a change in order to match the agreed-upon behavior.

I suspect we need to insert a step between steps 1 and 2 that says (hand-waving a bit): "if *element* has a computed value of the display property that is 'flex' or 'grid', then return the nearest ancestor containing block of the original element that you started with.

**Comment by Loirooriol** [comment]
Isn't it simpler to replace "if element is a not a block container" with "if element is a not a block box nor an inline-block"?
This excludes flex-level and grid-level block containers.

BTW, not sure if a table caption is considered to be a block box or a table-level block container:

```html
<div style="display: table; height:100px">
  <div style="display: table-caption">
    <div style="height: 100%; background: cyan">foobar
```
the inner `<div>` is 100px tall in Chromium and WebKit, but as tall as the text in Firefox.

**Comment by dholbert** [comment]
> Isn't it simpler to replace "if element is a not a block container" with "if element is a not a block box nor an inline-block"?
> This excludes flex-level and grid-level block containers.

I was originally thinking of something like that (amending step 4), but that won't help, because the algorithm can terminate before that and "choose" the flex container as the percent basis in step 3, if the flex container has a definite height.

I think:
(a) we need the algorithm to bail out *before step 3*, if it has landed on a grid/flex container (otherwise, step 3 could end up choosing that grid/flex container as the quirky percent-basis, to use for the descendant in question).
(b) we need to specify what element gets used as the containing block in that case.  (Right now the text says that you use the element that the algorithm returns, i.e. whatever ancestor you'd tunneled up to; but really, I think we want to use the nearest ancestor containing block for the percent-height element, or whatever we would've chosen in standards-mode.  And that nearest ancestor presumably will be indefinite height, which means the percent is unresolvable and treated as indefinite, which is OK.)

**Comment by Loirooriol** [comment]
> otherwise, step 3 could end up choosing that grid/flex container as the quirky percent-basis

But it would be chosen anyway at step 4 since it's not a block container? Isn't it fine? Why would we want to skip it if it has a definite height? Not sure if I get it, can you include some markup and which is the original element for which we are finding the containing block?

**Comment by dholbert** [comment]
Sorry, I think I misunderstood what you were suggesting -- you'd said "a block box", which I interpreted as another way of saying "a block container" / "a display:block box".

If you meant "a block-**level** box", then I think your suggested change would potentially work, because the flexbox and grid specs do indeed say that their flex/grid items are **not** block-level boxes (they're flex/grid-level).  So your suggested surgical change to the quirks-mode spec would effectively terminate the "digging up" algorithm as soon as we hit a flex item, and it would then choose the flex item as the percent basis (regardless of whether it has a definite size).

I'm still not sure it's useful to choose the flex item (rather than the percent-height-element's nearest containing block) as the percent basis in this scenario;  but maybe it doesn't matter if both are indefinite-height, as they probably are if we end up terminating the algorithm in this way.

**Comment by Loirooriol** [comment]
With [block box](https://drafts.csswg.org/css-display-3/#block-box) I meant "block-level block container", i.e. excluding flex/grid items (which are bot block-level) and excluding flex/grid containers (which are not block containers). Both need to be excluded because by using `position`, it can happen that the flex/grid item isn't in the containing block chain, but the flex/grid container is.

**Comment by dholbert** [comment]
Thanks - that makes sense. I had forgotten that "block box" was a specially defined term.

**Comment by Loirooriol** [comment]
Ping @dholbert for the spec update.

**Comment by dholbert** [comment]
Yup, sorry for going quiet here. There's a closely-related bug report that I want to better-understand & close out before fixing this - https://bugzilla.mozilla.org/show_bug.cgi?id=1687557 .  Hoping to take care of both soon.

**Comment by Loirooriol** [comment]
@dholbert Any updates?

**Comment by dholbert** [comment]
Sorry, only a little bit of an update; I have a patch posted on the bug that I mentioned (it was just an issue with one particular edge case), but the reviewer wants me to do a more thorough analysis of how this quirk applies with respect to orthogonal writing-modes (and how that matches/doesn't-match other browsers) before landing my patch, and I haven't had time for that yet.  And that analysis probably needs to be done before I write any spec text here, too.

(See https://bugzilla.mozilla.org/show_bug.cgi?id=1687557#c9 and subsequent testcases/comments from Mats in particular. the tl;dr, as I recall, is: Firefox applies this "percentage height quirk" **strictly for heights** (as you might expect), and Firefox bails out if the "walk up the tree" encounters any element that has a vertical writing mode. Whereas, Chromium seems to apply the quirk in the block axis, regardless of writing mode; i.e. it's more of a "percentage **block-size** quirk" in chromium.)

So: I need to do some more thorough investigation/test-crafting, and possibly raise some CSSWG discussion or file some Chrome bugs on the block-axis vertical-writing-mode stuff before I can usefully spec this (since that part was out of scope of the earlier csswg discussion/resolution).  Unfortunately I've had other tasks occupying my attention and haven't been able to get to that yet. (Also, on a personal note: I'll be out for the next ~3 weeks to move house, so that will be occupying most of my attention for the extremely-immediate future.)

**Comment by dholbert** [comment]
> Firefox applies this "percentage height quirk" **strictly for heights** (as you might expect), and Firefox bails out if the quirk's "walk up the tree" traverses any element that has a vertical writing mode. Whereas, Chromium seems to apply the quirk in the block axis, regardless of writing mode; i.e. it's more of a "percentage **block-size** quirk", in Chromium.)

For reference, I just filed https://bugzilla.mozilla.org/show_bug.cgi?id=1879957 on this behavior difference.

**Comment by zcorpan** [comment]
PR: https://github.com/whatwg/quirks/pull/76
